### PR TITLE
[Enhance] Support dynamic input shape for ViT-based algorithms

### PR DESCRIPTION
**Pull request by mzr1996**
## Motivation

Currently, our ViT-based algorithms need to specify the image shape. But in some downstream tasks, they need input images with various shapes.

## Modification

Refactor these ViT-based algorithms to support dynamic input shape as far as possible.

- :heavy_check_mark: VisionTransformer
- :heavy_check_mark: SwinTransformer
- :heavy_check_mark: DeiT
- :x:  MLP-Mixer (The structure doesn't support dynamic input shape)
- :heavy_check_mark: T2T-ViT

## BC-breaking (Optional)

The `input_resolution` and `auto_pad` argument are deprecated in the `ShiftWindowMSA`. And the `auto_pad` argument is removed from `SwinTransformer`.

## Use cases (Optional)

```python
>>> from mmcls.models import VisionTransformer
>>> import torch
>>> model = VisionTransformer(arch="base")
>>> inputs = torch.rand(1, 3, 256, 323)  # The input shape is no-limit.
>>> patch_embed, cls_token = model(inputs)[0]
>>> print(patch_embed.shape)
torch.Size([1, 768, 16, 21])
>>> print(cls_token.shape)
torch.Size([1, 768])
```

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/backbones/deit.py [hunk 1]
```diff
@@ -15,21 +15,38 @@ class DistilledVisionTransformer(VisionTransformer):
...
+            and 'deit-base'. If use dict, it should have below keys:
+
+            - **embed_dims** (int): The dimensions of embedding.
+            - **num_layers** (int): The number of transformer encoder layers.
+            - **num_heads** (int): The number of heads in attention modules.
+            - **feedforward_channels** (int): The hidden dimensions in
+              feedforward modules.
+
+            Defaults to 'deit-base'.
+        img_size (int | tuple): The expected input image shape. Because we
```
Comment (Ezra-Yu) [inline mmcls/models/backbones/deit.py:29]: can we di not set image_size and send a list irregular images

**Review by Ezra-Yu — COMMENTED** [review]
please add DeprecationWarning in PatchEmbed and PatchMerge.

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

File: mmcls/models/backbones/vision_transformer.py [hunk 2]
```diff
@@ -180,14 +195,17 @@ class VisionTransformer(BaseBackbone):
     num_extra_tokens = 1  # cls_token
 
     def __init__(self,
-                 arch='b',
+                 arch='base',
                  img_size=224,
                  patch_size=16,
+                 in_channels=3,
```
Comment (ZwwWayne) [inline mmcls/models/backbones/vision_transformer.py:201]: This introduces BC-breaking in mmselfsup 0.6.0.

**Comment by ShangWeiKuo** [comment]
Excuse me @mzr1996 . Could you please explain how did you do to make ViT-based algorithms to support dynamic input shape?

**Pull request merged**